src/Database/LSMTree/Internal/Vector/Growing.hs

```diff
@@ -0,0 +1,117 @@
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+
+{- HLINT ignore "Avoid restricted alias" -}
+
+-- | Vectors with support for appending elements.
+module Database.LSMTree.Internal.Vector.Growing
+(
+    GrowingVector,
+    new,
+    append,
+    unsafeFreeze
+)
+where
+
+import           Prelude hiding (init, last, length)
+
+import           Control.Monad (when)
+import           Control.Monad.ST.Strict (ST)
+import           Data.Primitive.PrimVar (PrimVar, newPrimVar, readPrimVar,
+                     writePrimVar)
+import           Data.STRef.Strict (STRef, newSTRef, readSTRef, writeSTRef)
+import           Data.Vector (Vector, force)
+import qualified Data.Vector as Mutable (unsafeFreeze)
+import           Data.Vector.Mutable (MVector)
+import qualified Data.Vector.Mutable as Mutable (grow, length, new, set, slice,
+                     take)
+
+{-|
+    A vector with support for appending elements.
+
+    Internally, the elements of a growing vector are stored in a buffer. This
+    buffer is automatically enlarged whenever this is needed for storing
+    additional elements. On each such enlargement, the size of the buffer is
+    multiplied by a power of 2, whose exponent is chosen just big enough to make
+    the final buffer size at least as high as the new vector length.
+-}
+data GrowingVector s a = GrowingVector
+                             !(STRef s (MVector s a)) -- Reference to the buffer
+                             !(PrimVar s Int)         -- Reference to the length
+
+-- | Creates a new, initially empty, vector.
+new :: Int                      -- ^ Initial buffer size
+    -> ST s (GrowingVector s a) -- ^ Construction of the vector
+new illegalInitialBufferSize | illegalInitialBufferSize <= 0
+    = error "Initial buffer size not positive"
+new initialBufferSize
+    = do
+        buffer <- Mutable.new initialBufferSize
+        bufferRef <- newSTRef buffer
```

Comment:
```suggestion
        bufferRef <- newSTRef $! buffer
```

Reply:
Hmm, I don’t think this is necessary. Since we’re using the strict version of `ST`, the computation `Mutable.new initialBufferSize` will definitely be performed before `newSTRef buffer`, and I’d be very surprised if `new` wouldn’t always yield a fully evaluated result. Thoughts?

Reply:
Okay. Fixed in d0cc37980fd55b9744549005340726080ca24a3a.